### DISCUSSION TITLE
Does `getStaticProps` run every push, or just when the file changes or something?

### DISCUSSION THREAD

**Discussion by lancejpollard** [vercel/next.js#47525] [Help]
### Summary

I have 35k+ static files (I was going to do a million, but maybe that's too much to ask at first).

```
info  - Generating static pages (0/34643)
--
06:49:40.882 | info  - Generating static pages (331/34643)
06:50:41.024 | info  - Generating static pages (657/34643)
06:51:41.404 | info  - Generating static pages (980/34643)
```

It's taken 20 minutes to get this far, so I assume I am going to run into the 45 minute build error. So what if my pages are static though? I am generating the pages based on 35k DB records, which I am loading from a JSON file. In the `getServerSideProps`, I was originally loading a 20mb file on every request, then selecting the dynamic record from that 35k record _map_. I decided it was a waste and somewhat slow to load such a large file on every request, then I saw `getStaticProps` and thought that may be the answer! But it appears like it is meant for just a few static pages, not thousands or millions, is that correct?

What do I do then in my case, what is the best approach?

I would think the best approach is to cache the file somehow, but I don't think that's possible on Next.js? I had trouble with "JSON modules" importing those (said "experimental support", and I ran into some weird blocker bugs), so can't do that even if it potentially caches them. Can I create a global variable somewhere and cache it like that?

Or what do you recommend? I don't want to add a database to my system, that would be costly and overkill.

**Comment by lancejpollard** [comment]
I just created a uti, which I import inside the `getServerSideProps` function, that seems like the right approach?

```ts
import fs from 'fs'
import getConfig from 'next/config'

const cache: Record<string, unknown> = {}

export default function loadJSON(path: string): unknown {
  if (cache[path]) {
    return cache[path]
  }

  // eslint-disable-next-line @typescript-eslint/no-unsafe-assignment
  const { serverRuntimeConfig } = getConfig()

  const data = JSON.parse(
    fs.readFileSync(
      // eslint-disable-next-line @typescript-eslint/no-unsafe-member-access
      `${serverRuntimeConfig.PROJECT_BASE}/${path}`,
      'utf-8',
    ),
  ) as unknown
  cache[path] = data
  return data
}
```

**Reply by lancejpollard** [reply]
Not 100% right, needed to call `path.resolve(staticPath)` to [hack it into the Vercel build](https://github.com/vercel/next.js/discussions/32236#discussioncomment-5427401)! Hopefully this feature doesn't just disappear 😄

**Reply by icyJoseph** [reply]
`process.cwd()` is the way to go though, https://nextjs.org/docs/api-reference/data-fetching/get-static-props#reading-files-use-processcwd

**Comment by icyJoseph** [comment]
You could use `getStaticProps` together with `getStaticPaths` to generate zero pages at build time (taking a extreme position here), but letting it have `fallback: true` (on getStaticPaths). Bots would use fallback blocking.

This way, you'd be forced to have logic for `router.isFallback`, which would be used when the user arrives to a page not-yet-generated. 

Pages would be generated as soon as a next/link points to them, unless prefetch is disabled. Careful with this, as it might a whole wave of page generation. If prefetch is disabled, hovering triggers the page creation.

In your case, since you statically have a JSON file as, sort of in memory DB, revalidation would be a waste. 

As long as you don't re-deploy, everything should remain statically generated IIRC. 

If you deploy (perhaps through git push hooks), IIRC, you'd be trashing down the generated stuff, and start over. 

- https://nextjs.org/docs/basic-features/data-fetching/incremental-static-regeneration#self-hosting-isr

In a sense, this is the cheapest, laziest way to go about it. Only pages that are potentially needed are generated. 

You could also say that you'd pre-render the top 5% most accessed pages, and do the rest as demanded. For those people with a direct link to a not yet generated page, you'd show the fallback page, while the page is generated, and then you'd immediately show content.
